feat: more prompt position strategies

l-kershaw opened the pull request:
This PR adds the `prompt_position` option for the following layout stratgies:
- `vertical`
- `bottom_pane`
- `center`

`vertical` was noted as being "planned" in the code
`bottom_pane` was requested in [this comment](https://github.com/nvim-telescope/telescope.nvim/issues/765#issuecomment-851753342)

---

EDIT: I have now also added some more flexibility to the `get_dropdown` and `get_ivy` themes if the user customises the `prompt_position`

Conni2461 commented on lua/telescope/pickers/layout_strategies.lua:
```diff
@@ -640,13 +640,32 @@ layout_strategies.vertical = make_documented_layout(
+        results.line = (preview.height == 0) and preview.line or preview.line + preview.height + (1 + bs)
+        prompt.line = results.line + results.height + (1 + bs)
+      else
+        error(
+          "Unknown prompt_position: " .. tostring(self.window.prompt_position) .. "\n" .. vim.inspect(layout_config)
+        )
```
I usually suggest `string.format`, it should make it a little bit more readable and you dont need `tostring`.

But isnt that big of a deal :laughing:

Conni2461 commented:
Just setting 

```lua
    layout_strategy = 'bottom_pane',
    layout_config = {
      bottom_pane = {
        prompt_position = "top" | "bottom",
      }
    }
```

looks a little bit of.
![image](https://user-images.githubusercontent.com/15233006/134811753-e85e1cf3-1eb8-40cc-a7f6-4035fec2fd37.png)
![image](https://user-images.githubusercontent.com/15233006/134811761-aee524bb-3be5-4689-977e-95c78d36feca.png)

FWIW plenary is up to date

l-kershaw commented:
I agree that it looks a bit weird without adjusting borderchars
I _think_ this design was a conscious choice by whoever made the `top` version, to then match the style of ivy for the theme.
I tried to match the general aesthetic for the `bottom` version

l-kershaw commented:
I've added some flexibility to the `get_dropdown` theme, so it will adjust based on `prompt_position`. Picture below:
![image](https://user-images.githubusercontent.com/35707277/134971787-83fb133d-b02a-439d-9d28-f2cf861c7e1a.png)

I've also added some flexibility to the `get_ivy` theme. Picture below:
![image](https://user-images.githubusercontent.com/35707277/134971628-ee594c2d-73f6-4fb8-a277-7679e824b9f0.png)

@Conni2461 is this more what you were looking for with the `prompt_position="bottom"` for the `bottom_pane` layout?

l-kershaw commented:
@fdschmidt93 I've rebased this PR as requested, and the fixes from #1293 still seem to work well.

If you could test this PR and let me know your thoughts, that would be great 🙂

Conni2461 reviewed (approved):
LGTM thanks :)

(just running layout_config `bottom_pane` without ivy layout still looks kinda weird, at least for me, but i dont think anyone is doing it. We can fix it if someone reports an issue for that)

pickfire commented:
`get_ivy` seemed to fill the whole screen. How should I configure the height?

l-kershaw commented:
@pickfire It depends on how you are using the telescope API.

---

If you are using the viml command line api, e.g.:
```viml
:Telescope find_files theme=get_ivy
```
then you could add `layout_config` to change the height by doing:
```viml
:Telescope find_files theme=get_ivy layout_config={height=10}
```

---

If you are using the lua api, e.g.:
```lua
require("telescope.builtin").find_files(require("telescope.themes").get_ivy())
```
then you can add `layout_config` to change the height by doing:
```lua
require("telescope.builtin").find_files(require("telescope.themes").get_ivy({layout_config={height=10}}))
```

---

If you want to create your own theme based on `get_ivy` then you can do the following in lua
```lua
require("telescope.themes").get_custom = function(opts)
  opts = vim.tbl_deep_extend("force",{layout_config={height=10}},opts)
  return require("telescope.themes").get_ivy(opts)
end
```
and then you can use you custom theme like:
```viml
:Telescope find_files theme=get_custom
```
or
```lua
require("telescope.builtin").find_files(require("telescope.themes").get_custom())
```

pickfire commented:
Wow, this thing is nested multiple layers deep, no wonder I couldn't find it in the docs.
